README
The University of Cambridge Web Authentication System Java Toolkit 
==================================================================

This toolkit provides assorted classes implementing functionality
associated with the University of Cambridge's Web Authentication
System (UcamWebAuth). UcamWebAuth is the software that implements the
University of Cambridge's 'Raven' system.

Current content of the package includes classes representing
authentication requests and responses (WebauthRequest and
WebauthResponse respectively), and a class for validating response
messages (WebauthValidator).


The Latest Version
------------------

See <http://raven.cam.ac.uk/project/java-toolkit/>


Installation
------------

A JAR file is provided containing the classes from this toolkit
ready-compiled. These can be made available to any Java program by
including the JAR file in the CLASSPATH. 

WebauthValidator needs access to a Java Cryptography Extension (JCE)
security provider that implements the "SHA1withRSA" signature
scheme. This appears to be the case by default from at least Sun's
J2SE 1.3.  Failing that, Sun's JCE reference implementation from
http://java.sun.com/products/jce/ or Bouncycastle's JCE implementation
from http://www.bouncycastle.org/ may be of use.


Documentation
-------------

Documentation is provided in the form of 'javadoc' files. Point a web
browser at javadoc/index.html.


Building
--------

The Maven tool <http://maven.apache.org/> is needed to rebuild the
toolkit. Assuming Maven is installed somewhere on PATH, the toolkit can
be rebuilt by executing the command

  maven compile 

from the root of the unpacked distribution. This will place the
freshly-compiled class files beneath the directory 'target/classes'. Other Maven
goals will run the provided JUnit tests, rebuild the JAR file and
rebuild the distribution jar files.


Questions
---------

Questions about the use of this toolkit should be addressed to
<raven-support@ucs.cam.ac.uk>. See also the pages at
<http://raven.cam.ac.uk/project/>


Licencing
---------

Copyright (c) University of Cambridge 2005,2014

This toolkit is free software; you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as
published by the Free Software Foundation; either version 2.1 of the
License, or (at your option) any later version. 

The toolkit is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this toolkit; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
USA

See the file COPYING for details.


CHANGES
University of Cambridge Web Authentication System Java Toolkit

                                 CHANGELOG

$Id: CHANGES,v 1.8 2007-03-18 19:04:11 jw35 Exp $

0.7-20141203 JW

* Added support for v3 of the Ucam WebAuth protocol:
    * Added basic v3 support
    * Modified validator rules - v1 url validation allows the response
      url if it is a prefix of the request url. v2 and up must match
      exactly.
    * Improved the code formatting, specifically the whitespacing with
      tabs replaced with spaces.

0.6-20051129 JW

* Made WebauthRequest Serializable since there is a good chance many
  applications will want to keep the object around to use to validate
  a coresponding WebauthResponse

0.5-20050728 JW

* Corrected a bug (similar to that in 0.4) that caused request
  timestamps to be interpreted as if in localtime not UT. Extended
  test suite to catch this

0.4-20050331 JW

* Corrected a bug that caused response timestamps to be interpreted as
  if in localtime not UT. Extended test suite to catch this

* Modified date format in error messages to be more readable and
  include time zone. 

0.3-20050330 JW

* Significant re-working, including numerous incompatible changes to
  the API (sorry). Generally:

  + The dedicated get...()/set...() methods from WebauthRequest and
    WebauthResponse have been replaced by generic methods that take
    the name of a field to act on as their first argument.

  + get() always returns a String, but there are convenience methods
    (getInt(), getDate() and getColl() ) that return fields as an int,
    a date, or as a Collection. These throw exceptions if the required
    convertion isn't possible.

  + set() is overloaded to accept a String, an int, a date, or a
    Collection as the value to set.

  + getColl() and set() with a Collection argument can be used to
    handle list-based fields (such as 'aauth' in requests and 'sso' in
    responses) as a Collection of Strings.

  + WebauthRequest and WebauthResponse now have methods that return the
    number of fields that they contain, and a iterator to access them
    all.

  + WebauthResponse no-longer implements java.lang.Principal. As a
    result toString() now returns a human-readable summary of the
    entire object. toFullString() is gone.

  + WebauthRequest and WebauthResponse now generally DON'T throw
    exceptions (except as mentioned above under convenience
    methods. As a result they both now do less validity checking then
    they did.

  + Dates are now uniformly represented as a long containing the
    number of milliseconds since January 1, 1970 GMT throughout.

* WebauthResponse.statusString() now accepts an int or a String
  argument.
 
* WebauthValidator checks that a status 200 response includes one of
  the 'auth' or 'sso' fields.

0.2-20050321 JW

* Added this file and TODO

* Added notes to README about the need for a security provider that
  implements the "SHA1withRSA" signature scheme, and the current use
  of sun.misc.BASE64Decoder

* Reworked (and commented!) timeout checking and get/set methods for
  Timeout and MaxSkew to cope with millisecond time resolution in Java
  but only second time resolution in response messages.

* Dropped use of URLEncoder.encode(String,String) in favour of
  'URLEncoder.encode(String)' since the former isn't present in
  1.3. Code now compiles and passes its tests in Java 1.3 (1.3.1_15),
  1.4 (1.4.2_07) and 5 (1.5.0_12)

* Added text from causal exception to "Failed to parse Webauth
  response issue date" and "Failed to parse Webauth response URL"
  messages. Can't use exception chaining becasue it's >=1.4 only.

* WebauthRequestTest and WebauthResponseTest reworked to use http:
  (rather than https:) URLs becasue 1.3 doesn't support https: out of
  the box.

* Moved the constants (STATUS_SUCCESS, etc) and the one remaining
  static method (statusString) from the Webauth class into
  WebauthResponse where they belong. Removed the Webauth class.

0.1-20050316 JW

* Initial release


src/test/java/uk/ac/cam/ucs/webauth/WebauthRequestTest.java
/* This file is part of the University of Cambridge Web Authentication
 * System Java Toolkit
 *
 * Copyright 2005,2014 University of Cambridge
 *
 * This toolkit is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * The toolkit is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this toolkit; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
 * USA
 *
 * $Id: WebauthRequestTest.java,v 1.4 2005/07/28 08:34:06 jw35 Exp $
 *
 */

package uk.ac.cam.ucs.webauth;

import java.io.FileOutputStream;
import java.io.ObjectOutputStream;
import java.util.Arrays;
import java.util.Date;
import java.util.HashSet;
import java.util.Iterator;

import junit.framework.TestCase;

public class WebauthRequestTest extends TestCase {

	private WebauthRequest request;

	// ----------------------------------------------------------- Housekeeping

	public static void main(String args[]) {
		junit.textui.TestRunner.run(WebauthRequestTest.class);
	}

	// --------------------------------------------------------------- Fixtures

	@Override
  protected void setUp() {

		request = new WebauthRequest();

		request.set("ver", 2);
		request.set("url", "http://www.cam.ac.uk/raven");
		request.set("desc", "Description");
		request.set("aauth", "x-foobar,pwd");
		request.set("msg", "Message");
		request.set("params", "Params");
		request.set("date", "20050305T123456Z");

	}

	@Override
  protected void tearDown() {
	}

	// ----------------------------------------------------------------- Tests

	public void testDefaults() throws java.text.ParseException {

		WebauthRequest r = new WebauthRequest();

		assertEquals("3", r.get("ver"));

		// the date in the ticket should be less than about a second
		// before now

		Date now = new Date();
		assertTrue(now.getTime() - r.getDate("date") >= 0);
		assertTrue(now.getTime() - r.getDate("date") < 1000);

	}

	public void testLength() {
		assertEquals(7, request.length());
	}

	public void testGetFieldNames() {
		String[] e = { "ver", "url", "desc", "aauth", "msg", "params", "date" };
		HashSet<String> expected = new HashSet<String>(Arrays.asList(e));
		HashSet<String> got = new HashSet<String>();
		for (Iterator<String> it = request.getFieldNames(); it.hasNext();)
			got.add(it.next());
		assertEquals(expected, got);
	}

	public void testToQString() {

		// Check that request.toQString is as expected

		assertEquals("ver=2&url=http%3A%2F%2Fwww.cam.ac.uk%2Fraven"
				+ "&desc=Description&aauth=x-foobar%2Cpwd&msg=Message"
				+ "&params=Params&date=20050305T123456Z", request.toQString());

		// now additionally set fail to true and test again

		request.set("fail", "yes");

		// System.out.println(request.toQString());
		assertEquals("ver=2&url=http%3A%2F%2Fwww.cam.ac.uk%2Fraven"
				+ "&desc=Description&aauth=x-foobar%2Cpwd&msg=Message"
				+ "&params=Params&date=20050305T123456Z&fail=yes",
				request.toQString());

		// and the same for 'iact' set to "no"

		request.set("iact", "no");

		// System.out.println(request.toQString());
		assertEquals("ver=2&url=http%3A%2F%2Fwww.cam.ac.uk%2Fraven"
				+ "&desc=Description&aauth=x-foobar%2Cpwd&iact=no"
				+ "&msg=Message&params=Params&date=20050305T123456Z"
				+ "&fail=yes", request.toQString());

		// and finally for 'iact' set to "yes"

		request.set("iact", "yes");

		// System.out.println(request.toQString());
		assertEquals("ver=2&url=http%3A%2F%2Fwww.cam.ac.uk%2Fraven"
				+ "&desc=Description&aauth=x-foobar%2Cpwd&iact=yes"
				+ "&msg=Message&params=Params&date=20050305T123456Z"
				+ "&fail=yes", request.toQString());

	}

	public void testToString() {

		assertEquals("Webauth request: ver: 2, url: http://www.cam.ac.uk/"
				+ "raven, desc: Description, aauth: x-foobar,pwd, iact"
				+ ": , msg: Message, params: Params, date: 20050305T12"
				+ "3456Z, fail: ", request.toString());

	}

	public void testGet() {
		assertEquals("2", request.get("ver"));
	}

	public void testGetInt() {

		assertEquals(2, request.getInt("ver"));

		try {
			request.getInt("url");
			fail("didn't spot non-integer");
		} catch (NumberFormatException e) {
			assertTrue(true);
		}

	}

	public void testGetDate() throws java.text.ParseException {

		assertEquals(1110026096000L, request.getDate("date"));

		// Try again with a date in the summer

		WebauthRequest r = new WebauthRequest();
		r.set("date", "20050605T123456Z");
		assertEquals(1117974896000L, r.getDate("date"));

		try {
			request.getDate("url");
			fail("didn't spot non-date");
		} catch (java.text.ParseException e) {
			assertEquals("java.text.ParseException: Unparseable date: "
					+ "\"http://www.cam.ac.uk/raven\"", e.toString());
		}

	}

	public void testGetColl() {

		HashSet<String> set = new HashSet<String>();
		set.add("x-foobar");
		set.add("pwd");
		assertEquals(set, request.getColl("aauth"));

		HashSet<String> set2 = new HashSet<String>();
		set2.add("2");
		assertEquals(set2, request.getColl("ver"));

	}

	public void testGetNonsuch() {
		assertEquals("", request.get("nonsuch"));
		assertEquals(-1, request.getInt("nonsuch"));

		HashSet<String> set = new HashSet<String>();
		assertEquals(set, request.getColl("nonsuch"));

		try {
			request.getDate("nonsuch");
			fail("didn't spot non-existent date");
		} catch (java.text.ParseException e) {
			assertEquals("java.text.ParseException: Unparseable date: \"\"",
					e.toString());
		}
	}

	public void testGetNull() {
		assertEquals("", request.get(null));
		assertEquals(-1, request.getInt(null));

		HashSet<String> set = new HashSet<String>();
		assertEquals(set, request.getColl(null));

		try {
			request.getDate(null);
			fail("didn't spot non-existent date");
		} catch (java.text.ParseException e) {
			assertEquals("java.text.ParseException: Unparseable date: \"\"",
					e.toString());
		}
	}

	public void testSet() {

		request.set("foo", "bar");
		assertEquals("bar", request.get("foo"));

		request.set("foo", 999);
		assertEquals("999", request.get("foo"));

		request.set("foo", 1110026096000L);
		assertEquals("20050305T123456Z", request.get("foo"));

		request.set("foo", 1117974896000L);
		assertEquals("20050605T123456Z", request.get("foo"));

		HashSet<String> set = new HashSet<String>();
		set.add("x-foobar");
		set.add("pwd");
		request.set("foo", set);
		assertEquals("x-foobar,pwd", request.get("foo"));

	}

	public void testSerialization() throws java.io.FileNotFoundException,
			java.io.IOException {

		FileOutputStream fileOut = new FileOutputStream(getClass().getResource(
				"/").getFile()
				+ "ser.test");
		ObjectOutputStream out = new ObjectOutputStream(fileOut);
		out.writeObject(request);
		out.close();
		assertTrue(true);
	}

}


src/main/java/uk/ac/cam/ucs/webauth/Util.java
/* This file is part of the University of Cambridge Web Authentication
 * System Java Toolkit
 *
 * Copyright 2012 University of Cambridge
 *
 * This toolkit is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * The toolkit is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this toolkit; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
 * USA
 *
 */
package uk.ac.cam.ucs.webauth;

/** private utility methods */
public class Util {

	/*
	 * Split a string on delim. Note this has different semantics than the java
	 * standard library split
	 */
	static String[] split(char delim, String text) {

		if (text.length() == 0)
			return new String[0];

		int nFields = 0;
		for (int pos = 0; pos < text.length(); ++pos)
			if (text.charAt(pos) == delim)
				++nFields;
		String[] list = new String[nFields + 1];

		int toIndex = 0;
		int fromIndex = 0;
		int ctr = 0;

		while (fromIndex <= text.length()) {
			toIndex = text.indexOf(delim, fromIndex);
			if (toIndex == -1)
				toIndex = text.length();
			list[ctr++] = text.substring(fromIndex, toIndex);
			fromIndex = toIndex + 1;
		}

		return list;

	}
}


src/main/java/uk/ac/cam/ucs/webauth/WebauthDecoder.java
/* This file is part of the University of Cambridge Web Authentication
 * System Java Toolkit
 *
 * Copyright 2005,2014 University of Cambridge
 *
 * This toolkit is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * The toolkit is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this toolkit; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
 * USA
 *
 * $Id: WebauthDecoder.java,v 1.3 2005/03/30 13:17:05 jw35 Exp $
 *
 */

package uk.ac.cam.ucs.webauth;

import java.io.IOException;

import org.apache.commons.codec.binary.Base64;

/**
 * Implements a decoder for encoded binary strings as used by the
 * authentication system. The encoding scheme is actually just BASE64
 * as defined in RFC 1521, except that the characters '+', '/', and '='
 * are replaced by '-', '.' and '_' to reduce the URL-encoding
 * overhead.
 *
 * @version $Revision: 1.3 $ $Date: 2005/03/30 13:17:05 $
 */

public class WebauthDecoder {

    /** 
     * Default constructor
     *
     */
     
    public WebauthDecoder() {
    }

    /** 
     * Decode a string.
     *
     * @param encoded  the encoded string for processing
     *
     * @return the decoded data
     *
     * @throws IOException if the encoded string is malformed 
     */

    public byte[] decodeBuffer(String encoded) throws IOException {
	
	StringBuffer buff = new StringBuffer(encoded.length());
	for (int i=0; i<encoded.length();++i) {
	    char c = encoded.charAt(i);
	    if (c == '-') {
		c = '+';
	    }
	    else if (c == '.') {
		c = '/';
	    }
	    else if (c == '_') {
		c = '=';
	    }
	    buff.append(c);
	}

	return Base64.decodeBase64(buff.toString().getBytes());
    }
 
}



src/main/java/uk/ac/cam/ucs/webauth/WebauthValidator.java
/* This file is part of the University of Cambridge Web Authentication
 * System Java Toolkit
 *
 * Copyright 2005,2014 University of Cambridge
 *
 * This toolkit is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * The toolkit is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this toolkit; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
 * USA
 *
 * $Id: WebauthValidator.java,v 1.12 2005/03/31 15:10:07 jw35 Exp $
 *
 */

package uk.ac.cam.ucs.webauth;

import java.io.IOException;
import java.security.InvalidKeyException;
import java.security.KeyStore;
import java.security.KeyStoreException;
import java.security.NoSuchAlgorithmException;
import java.security.Signature;
import java.security.SignatureException;
import java.security.cert.Certificate;
import java.text.SimpleDateFormat;
import java.util.Collection;
import java.util.Date;

/**
 * Implements a validator for authentication response message.
 * 
 * @version $Revision: 1.12 $ $Date: 2005/03/31 15:10:07 $
 */

public class WebauthValidator {

	private static final int MAX_VER = 3;
	private static final String SIGNATURE_SCHEME = "SHA1withRSA";

	private static final int DEFAULT_TIMEOUT = 30000;
	private static final int DEFAULT_MAX_SKEW = 500;
	private static final String DEFAULT_KEY_PREFIX = "webauth-pubkey";

	private static final String DATE_FORMAT = "yyyy-MM-dd HH:mm:ss.SSS z";

	private KeyStore keyStore;
	private int timeout = DEFAULT_TIMEOUT;
	private int maxSkew = DEFAULT_MAX_SKEW;
	private String keyPrefix = DEFAULT_KEY_PREFIX;

	/**
	 * Default constructor. The timeout for the resulting object is set to 30
	 * sec, the clock skew to 0.5s, and the key prefix to "webauth-pubkey".
	 * 
	 * <p>
	 * 
	 * The validator needs access to the public key (or occasionally keys)
	 * corresponding to the private key(s) being used by the WLS to sign
	 * responses. These keys must be available in certificates contained in the
	 * {@link java.security.KeyStore} passed to the constructor. These should be
	 * stored under aliases formed by concatenating the key prefix and the KeyID
	 * used to identify a key in a response - for example by default the key
	 * identified in a response by '2' should be stored under the alias
	 * 'webauth-pubkey2'. Note that for the correct operation of the system it
	 * is important that exactly the right set of keys are provided to the
	 * validator - new keys should be added before they start to be used by the
	 * WLS, and old keys MUST be removed once they are no longer in use since
	 * otherwise if compromised they would represent a security vulnerability.
	 * 
	 * <p>
	 * 
	 * Any KeyStore implementing the standard Java Crypto API can be used for
	 * this. The most common is the "JKS" store implemented in the "SUN" crypto
	 * provider. Such a KeyStore can either be initialised from a keystore file
	 * on disk maintained by the <tt>keytool</tt> utility, or can be initialised
	 * at run-time with certificates stored elsewhere. For example:
	 * 
	 * <pre>
	 * import java.security.KeyStore;
	 * import java.security.cert.Certificate;
	 * import java.security.cert.CertificateFactory;
	 * 
	 * KeyStore ks = KeyStore.getInstance("JKS");
	 * ks.load(null, new char[] {}); // Null InputStream, no password     
	 * 
	 * CertificateFactory factory = 
	 *     CertificateFactory.getInstance("X.509");
	 * Certificate cert = 
	 *     factory.generateCertificate(new FileInputStream("pubkey2.crt"));
	 * ks.setCertificateEntry("webauth-pubkey2",cert);
	 * </pre>
	 * 
	 * @param k
	 *            A {@link java.security.KeyStore} containing the
	 *            currently-valid public keys for the authentication system.
	 */

	public WebauthValidator(KeyStore k) {
		keyStore = k;
	}

	/**
	 * Perform validation tests on a WebauthResponse. This involves:
	 * <ol>
	 * <li>Checking that an acceptable combination of parameters are present in
	 * the response.
	 * 
	 * <li>Checking that 'kid', if present, corresponds to a key currently being
	 * used by the WAA.
	 * 
	 * <li>Checking that the signature, if provided, matches the data supplied.
	 * 
	 * <li>Checking that the response is recent by comparing 'issue' with the
	 * supplied date. If the supplied date is not from a clock synchronised by
	 * NTP or a similar mechanism then an allowance must be made for the maximum
	 * expected clock skew.
	 * 
	 * <li>Checking that 'url' is consistent with that in the corresponding
	 * Request.
	 * 
	 * <li>Checking that 'auth' and/or 'sso' contain values that are consistent
	 * with those in the corresponding Requestg.
	 * </ol>
	 * 
	 * @param request
	 *            The {@link WebauthRequest WebauthRequest} object, the
	 *            submission of which to the login server resulted in the
	 *            WebauthResponse being validated. This is not required to be
	 *            the identical object, but it should contain the same values
	 *            for the following parameters as the object which did (or at
	 *            least could) have caused this response: ver, url, iact, aauth.
	 *            Note that the url in the request is only required to be a
	 *            prefix of the url in the response.
	 * 
	 * @param response
	 *            The {@link WebauthResponse WebauthResponse} object to be
	 *            validated
	 * 
	 * @throws WebauthException
	 *             if the response fails to validate
	 */

	public void validate(WebauthRequest request, WebauthResponse response)
			throws WebauthException {

		validate(request, response, new Date().getTime());

	}

	/**
	 * Alternate version of {@link #validate(WebauthRequest, WebauthResponse)
	 * validate} in which date on which validation is based can be specified.
	 * 
	 * @param request
	 *            See {@link #validate(WebauthRequest, WebauthResponse)}
	 * @param response
	 *            See {@link #validate(WebauthRequest, WebauthResponse)}
	 * @param date
	 *            The date on which validation should be based, expressed as the
	 *            number of milliseconds since January 1, 1970 GMT..
	 * 
	 * @throws WebauthException
	 *             if the response fails to validate
	 */

	public void validate(WebauthRequest request, WebauthResponse response,
			long date) throws WebauthException {

	  if (null == request) {
	    throw new WebauthException("Missing WebauthRequest information");
	  }
	  if (null == response) {
	    throw new WebauthException("Missing WebauthResponse information");
	  }
    check_parameters(response);
    check_status(response);
    check_time(response, date);
    check_sig(response);

    check_protocol(request, response);
    check_url(request, response);
    check_iact(request, response);
    check_aauth(request, response);
	}

	private void check_parameters(WebauthResponse response)
			throws WebauthException {

		if (response.get("ver").equals(""))
			throw new WebauthException("Protocol version number missing "
					+ "from response");

		if (response.get("status").equals(""))
			throw new WebauthException("Status code missing from response");

		if (response.get("issue").equals(""))
			throw new WebauthException("Issue date missing from response");

		if (response.get("id").equals(""))
			throw new WebauthException("Response ID missing");

		if (response.get("url").equals(""))
			throw new WebauthException("URL missing from response");

		if (response.get("principal").equals("")
				&& response.getInt("status") == WebauthResponse.SUCCESS)
			throw new WebauthException("Principal missing from "
					+ "status 200 response");

		if (response.get("auth").equals("") && response.get("sso").equals("")
				&& response.getInt("status") == WebauthResponse.SUCCESS)
			throw new WebauthException("No authentication type "
					+ "found in status 200 response");

		if (!response.get("auth").equals("") && !response.get("sso").equals(""))
			throw new WebauthException("Both first-hand and SSO auth tokens "
					+ "found in response");

		if ((response.get("kid").equals("") || response.get("sig").equals(""))
				&& response.getInt("status") == WebauthResponse.SUCCESS)
			throw new WebauthException("KeyID and/or signature missing from "
					+ "status 200 response");
	}

	private void check_protocol(WebauthRequest request, WebauthResponse response)
			throws WebauthException {

		if ((response.getInt("ver") > MAX_VER)
				|| (response.getInt("ver") > request.getInt("ver"))) {
			throw new WebauthException("Unacceptable protocol version ("
					+ response.get("ver") + ") in response");
		}
	}

	private void check_url(WebauthRequest request, WebauthResponse response)
			throws WebauthException {

		String requestURL = request.get("url");

    // The url sent back inside the WLS-Response.
    String responseURL = response.get("url");
    // The responseURL is stored in a decoded state which will have converted + into a space.
    // Lets also do this for the comparison.
    //
    // >> Ideally we would have written a Raven specific decoder based on a modified copy of
    // >> java.net.ULRDecoder but the licensing might be a problem.
    requestURL = requestURL.replace("+", " ");

    if ("1".equals(response.get("ver"))) {
      // Version 1 protocol responds without the query string.

      // remove any query strings for comparison
      requestURL = requestURL.replaceFirst("[?].*", "");
      responseURL = responseURL.replaceFirst("[?].*", "");

      if (!responseURL.startsWith(requestURL)) {
        throw new WebauthException("URL in response (" + responseURL
            + ") does not match expected URL (" + requestURL + ")");
      }
    } else {
      // Version 2+. Do any exact match on the url.
      if (!requestURL.equals(responseURL)) {
        throw new WebauthException("URL in response (" + responseURL
            + ") does not match expected URL (" + requestURL + ")");
      }

    }

  }

	private void check_status(WebauthResponse response) throws WebauthException {

		int status = response.getInt("status");
		if (status != WebauthResponse.SUCCESS) {
			String msg = WebauthResponse.statusString(status);
			if (!response.get("msg").equals("")) {
				msg = msg + ": " + response.get("msg");
			}
			throw new WebauthException(msg);
		}

	}

	private void check_time(WebauthResponse response, long now)
			throws WebauthException {

		long issue = response.getDate("issue");

		// We require Issue <= Now + Skew

		if (issue > now + maxSkew) {
			SimpleDateFormat formater = new SimpleDateFormat(DATE_FORMAT);
			throw new WebauthException("Response apparently issued "
					+ "in the future; " + "issue time "
					+ formater.format(new Date(issue)) + " compared with "
					+ formater.format(new Date(now)));
		}

		// ... and that Now - Skew <= Issue + Timeout

		// Note that the actual moment of issue could have been up to
		// 1s after the issue time from the response message because
		// the latter only has second resolution and is (probably)
		// rounded down to the nearest second. As a result the timeout
		// applied could be upto 1s less than the value of
		// timeout. Since timeouts of less than about 30s are a bad
		// idea for other reasons it does not seem worth allowig for
		// this here.

		if (now - maxSkew > issue + timeout) {
			SimpleDateFormat formater = new SimpleDateFormat(DATE_FORMAT);
			throw new WebauthException("Response issued too long ago; "
					+ "issue time " + formater.format(new Date(issue))
					+ " compared with " + formater.format(new Date(now)));
		}

	}

	private void check_iact(WebauthRequest request, WebauthResponse response)
			throws WebauthException {

		if (request.get("iact").equalsIgnoreCase("yes")
				&& response.get("auth").equals("")) {
			throw new WebauthException("First-hand authentication required "
					+ "but not supplied");
		}

	}

	private void check_aauth(WebauthRequest request, WebauthResponse response)
			throws WebauthException {

		Collection<String> aauth = request.getColl("aauth");

		if (aauth == null || aauth.isEmpty())
			return;

		String auth = response.get("auth");
		Collection<String> sso = response.getColl("sso");
		for (String next : aauth) {
			if (auth.equals(next) || sso.contains(next)) {
				return;
			}
		}
		throw new WebauthException("No acceptable authentication types used");
	}

	private void check_sig(WebauthResponse response) throws WebauthException {

		try {
			Certificate cert = keyStore.getCertificate(keyPrefix
					+ response.get("kid"));
			if (cert == null) {
				throw new WebauthException("Failed to retrieve a key with "
						+ "alias " + keyPrefix + response.get("kid")
						+ " from the key store");
			}

			WebauthDecoder decoder = new WebauthDecoder();
			byte[] sigBytes = decoder.decodeBuffer(response.get("sig"));

			Signature signature = Signature.getInstance(SIGNATURE_SCHEME);
			signature.initVerify(cert);
			signature.update(response.getRawData().getBytes());

			if (!signature.verify(sigBytes)) {
				throw new WebauthException(
						"Unable to verify response signature");
			}
		} catch (KeyStoreException e) {
			throw new WebauthException("Validator keyStore object "
					+ "not correctly initialized");
		} catch (IOException e) {
			throw new WebauthException("Failed to decode signature string");
		} catch (NoSuchAlgorithmException e) {
			throw new WebauthException("No security provider implementing "
					+ "signature scheme " + SIGNATURE_SCHEME
					+ " available in this VM");
		} catch (InvalidKeyException e) {
			throw new WebauthException("Key with alias " + keyPrefix
					+ response.get("kid") + " in the key store is invalid");
		} catch (SignatureException e) {
			throw new WebauthException("Failed to verify signature - "
					+ "signature object is not initialized");
		}

	}

	/**
	 * Set the maximum expected transmission time for response messages, in
	 * milliseconds. A response recieved more than this time after it was issued
	 * will be considered invalid. The default is 30s, and since in some
	 * situations response transmition may require user intervention (for
	 * example to confirm a redirect) it should probably not be set much lower
	 * than this. However in an environment subject to network snooping this
	 * should be kept as short as realistically possible.
	 * 
	 * @param timeout
	 *            the integer value to which timeout should be set.
	 */

	public void setTimeout(int timeout) {
		this.timeout = timeout;
	}

	/**
	 * Get the maximum expected transmission time for response messages, in
	 * milliseconds. See {@link #setTimeout setTimeout} for details.
	 * 
	 * @return the integer value representing the timeout in milliseconds
	 */

	public int getTimeout() {
		return timeout;
	}

	/**
	 * Set the maximum expected difference between the clock supplying the date
	 * parameter for {@link #validate validate} and correct time, in
	 * milliseconds. The default is 0.5s. Given that exact clock synchronisation
	 * is near to impossible it should probably not be set significantly lower
	 * than this. In particular, at very low values there is likelyhood that
	 * even minor clock skew and/or rounding error will result in responses
	 * being rejected becasue the y appear to have been issued in the future.
	 * 
	 * @param maxSkew
	 *            the integer value of the maximum expected clock skew.
	 */

	public void setMaxSkew(int maxSkew) {
		this.maxSkew = maxSkew;
	}

	/**
	 * Get the maximum expected clock skew. See {@link #setMaxSkew setMaxSkew}
	 * for details
	 * 
	 * @return the integer value representing the maximum expected clock skew.
	 */

	public int getMaxSkew() {
		return maxSkew;
	}

	/**
	 * Set the string prefix used to identify the relevant public key in the key
	 * store. Keys must be available in the key store under an alias formed from
	 * this prefix and the key-id carried in the response message being
	 * validated.
	 * 
	 * @param keyPrefix
	 *            the prefix string
	 */

	public void setKeyPrefix(String keyPrefix) {
		this.keyPrefix = keyPrefix;
	}

	/**
	 * Get the maximum expected clock skew. See {@link #setKeyPrefix
	 * setKeyPrefix} for details
	 * 
	 * @return the prefix string
	 */

	public String getKeyPrefix() {
		return keyPrefix;
	}

}
